finality-aleph/src/data_io.rs

```diff
@@ -1,22 +1,214 @@
+                Some(message) = &mut self.messages_rx.next() => {
+                    trace!(target: "afa", "Received message at Data Store {:?}", message);
+                    self.add_message(message);
+                }
+                Some(block) = &mut import_stream.next() => {
+                    trace!(target: "afa", "Block import notification at Data Store for block {:?}", block);
+                    self.add_block(block.hash);
+                }
+                _ = &mut maintenance_timeout => {
+                    trace!(target: "afa", "Data Store maintenance timeout");
+                    let keys : Vec<_> = self.dependent_messages.keys().cloned().collect();
+                    for block_hash in keys {
+                        if let Ok(Some(_)) = self.client.header(BlockId::Hash(block_hash)) {
+                            self.add_block(block_hash);
+                        }
+                    }
+                    maintenance_timeout = Delay::new(PERIODIC_MAINTENANCE_INTERVAL);
+                }
+                _ = &mut exit => {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn forget_message(&mut self, message_id: MessageId) {
+        self.message_requirements.remove(&message_id);
+        if let Some(message) = self.pending_messages.remove(&message_id) {
+            for block_hash in message.included_blocks() {
+                if let Entry::Occupied(mut entry) = self.dependent_messages.entry(block_hash) {
+                    entry.get_mut().remove(&message_id);
+                    if entry.get().is_empty() {
+                        entry.remove_entry();
+                    }
+                }
+            }
+        }
+    }
+
+    fn add_pending_message(&mut self, message: Message, requirements: Vec<B::Hash>) {
+        let message_id = self.next_message_id;
+        self.next_message_id += 1;
+        for block_hash in requirements.iter() {
+            self.dependent_messages
+                .entry(*block_hash)
+                .or_insert_with(HashSet::new)
+                .insert(message_id);
+        }
+        self.message_requirements
+            .insert(message_id, requirements.len());
+        self.pending_messages.insert(message_id, message);
+
+        if message_id >= MESSAGE_ID_BOUNDARY {
+            self.forget_message(message_id - MESSAGE_ID_BOUNDARY)
+        }
+    }
+
+    fn add_message(&mut self, message: Message) {
+        let requirements: Vec<_> = message
+            .included_blocks()
+            .into_iter()
+            .filter(|block_hash| {
+                if self.available_blocks.contains(block_hash) {
+                    return false;
+                }
+                if let Ok(Some(_)) = self.client.header(BlockId::Hash(*block_hash)) {
+                    self.add_block(*block_hash);
+                    return false;
+                }
+                true
+            })
+            .collect();
```

Comment:
You don't store anywhere requirements as a Vector, so you could leave it as Iterator.

Reply:
After that I am passing requirements to `self.add_pending_messages(message, requirements)` and also checking if it is empty and checking its size, so collect is more convenient in use here. Otherwise I would have to use `.peekable()` or something like that here.

Reply:
Vector allocates so sometimes convenience is not justified. In this case, it seems ok though.